### PR TITLE
Button update

### PR DESCRIPTION
**Pull request by MrTibbles**
- Text Button
- Text + Icon Button (different positions)
- Borderless Text Button

### CODE REVIEW

File: lib/Button/Button.js [hunk 1]
```diff
@@ -17,14 +17,34 @@ const StyledButton = styled.button.attrs({
...
   ${props => size(...props.size)};
   background-color: ${({ backgroundColor }) => backgroundColor};
+  background: ${({ icon }) => icon && `url(${icon}) no-repeat`};
+  background-position: ${({ iconPos }) => iconPos};
+  background-size: ${({ iconSize }) => iconSize};
   border-color: ${({ borderColor }) => borderColor};
   border-style: ${({ borderStyle }) => borderStyle};
   border-width: ${({ borderWidth }) => borderWidth};
   border-radius: ${({ borderRadius }) => borderRadius};
+  border: ${({ border }) => !border && 'none'};
```
Comment (lukehedger) [inline lib/Button/Button.js]: Can just use `borderWidth={0}` here, instead of shorthand `border` prop.

File: lib/Button/Button.js [hunk 2]
```diff
@@ -17,14 +17,34 @@ const StyledButton = styled.button.attrs({
   ${props => padding(...props.padding)};
   ${props => size(...props.size)};
   background-color: ${({ backgroundColor }) => backgroundColor};
+  background: ${({ icon }) => icon && `url(${icon}) no-repeat`};
+  background-position: ${({ iconPos }) => iconPos};
+  background-size: ${({ iconSize }) => iconSize};
```
Comment (lukehedger) [inline lib/Button/Button.js]: Use `backgroundSize` as prop name to keep it predictable.

File: lib/Button/Button.js [hunk 3]
```diff
@@ -17,14 +17,34 @@ const StyledButton = styled.button.attrs({
   ${props => padding(...props.padding)};
   ${props => size(...props.size)};
   background-color: ${({ backgroundColor }) => backgroundColor};
+  background: ${({ icon }) => icon && `url(${icon}) no-repeat`};
+  background-position: ${({ iconPos }) => iconPos};
```
Comment (lukehedger) [inline lib/Button/Button.js]: Use `backgroundPosition` as prop name to keep it predictable.

File: lib/Button/Button.js [hunk 4]
```diff
@@ -17,14 +17,34 @@ const StyledButton = styled.button.attrs({
   ${props => padding(...props.padding)};
   ${props => size(...props.size)};
   background-color: ${({ backgroundColor }) => backgroundColor};
+  background: ${({ icon }) => icon && `url(${icon}) no-repeat`};
```
Comment (lukehedger) [inline lib/Button/Button.js]: Split this into `backgroundImage` (with `src` as prop name) and `backgroundRepeat` (with `no-repeat` as default value).

File: lib/Button/Button.js [hunk 5]
```diff
@@ -17,14 +17,34 @@ const StyledButton = styled.button.attrs({
...
   border-color: ${({ borderColor }) => borderColor};
   border-style: ${({ borderStyle }) => borderStyle};
   border-width: ${({ borderWidth }) => borderWidth};
   border-radius: ${({ borderRadius }) => borderRadius};
+  border: ${({ border }) => !border && 'none'};
   color: ${({ color }) => color};
   cursor: pointer;
   outline: none;
 
+  transition: color 250ms ease-in, background-color 250ms ease-in;
```
Comment (lukehedger) [inline lib/Button/Button.js]: Should be able to set this with prop.

File: lib/Button/Button.js [hunk 6]
```diff
@@ -17,14 +17,34 @@ const StyledButton = styled.button.attrs({
...
+
+  &:hover {
```
Comment (lukehedger) [inline lib/Button/Button.js]: Would be good to make this as flexible as poss, without needing to extend `switch`. Could pass this as a `hover` object that is defined on the theme?

Something like...
```
&:hover {
  {({ hover }) => ...hover}
}
```
`theme.js`
```
const theme = {
  hover: {
    background: {
      backgroundColor: '',
      color: '',
    },
    ...
  }
  ...
}
```

File: lib/Button/Button.js (hunk 2 above)
Comment (MrTibbles) [reply lib/Button/Button.js]: At the time `icon` made sense, looking back at it now, it makes very little!

File: lib/Button/Button.js (hunk 1 above)
Comment (MrTibbles) [reply lib/Button/Button.js]: It needs to be `none` to remove default styling, much like `outline` value. However it should have a more semantic name, `noBorder` for example.

File: lib/Button/Button.js (hunk 4 above)
Comment (MrTibbles) [reply lib/Button/Button.js]: I did it like that to save repeating the same conditional check on `icon` being defined, felt neater. But for clarity and consistency i agree.

Comment (lukehedger) [reply lib/Button/Button.js]: Ah I see what you mean. The default can save us having to specify `backgroundRepeat`

File: lib/Button/Button.js (hunk 1 above)
Comment (lukehedger) [reply lib/Button/Button.js]: `borderWidth={0}` will achieve this. Ideally, we are only passing valid CSS as props to minimise overhead.

File: lib/Button/Button.js (hunk 6 above)
Comment (lukehedger) [reply lib/Button/Button.js]: Another approach is to extend a primitive into a custom component used at application-level:
```
const MyButton = styled(Button)`
  &:hover {
    ...
  }
`
```

File: lib/Button/Button.js [hunk 7]
```diff
@@ -17,32 +17,22 @@ const StyledButton = styled.button.attrs({
...
+  background-image: ${({ backgroundImage }) => `url(${backgroundImage})`};
+  background-repeat: ${({ backgroundRepeat }) => backgroundRepeat};
+  background-position: ${({ backgroundPos }) => backgroundPos};
+  background-size: ${({ backgroundSize }) => backgroundSize};
   border-color: ${({ borderColor }) => borderColor};
   border-style: ${({ borderStyle }) => borderStyle};
   border-width: ${({ borderWidth }) => borderWidth};
   border-radius: ${({ borderRadius }) => borderRadius};
-  border: ${({ border }) => !border && 'none'};
+  border: ${({ noBorder }) => noBorder && 'none'};
```
Comment (lukehedger) [inline lib/Button/Button.js]: borderWidth={0} will achieve this. Ideally, we are only passing valid CSS as props to minimise overhead.

Comment (lukehedger) [reply lib/Button/Button.js]: Not sure we need a border shorthand prop?

**Pull request merged**